### PR TITLE
MLHR-1938 - Operator checkpointing in distributed in-memory store

### PR DESCRIPTION
**Pull request by ashishtadose**
Depends on https://issues.apache.org/jira/browse/APEXCORE-283

@tushargosavi kindly review.

### CODE REVIEW

File: library/src/main/java/com/datatorrent/lib/util/AppIdAwareStorageAgent.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+package com.datatorrent.lib.util;
+
+import com.datatorrent.api.StorageAgent;
+
+/**
+ * Interface to pass application id to storage agent
+ * 
+ *
+ */
+public interface AppIdAwareStorageAgent extends StorageAgent
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/AppIdAwareStorageAgent.java]: Is this required here? Will ApplicationAwareStorageAgent defined in Apex not take care of this?

File: library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+package com.datatorrent.lib.util;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Interface for KeyValue store
+ * 
+ */
+public interface StorageAgentKeyValueStore
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java]: There is com.datatorrent.lib.db.KeyValueStore in malhar. Can this not be used?

File: library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+   * @return the list of all associated keys
+   */
+  public List<String> getKeys(Object key);
+
+  /**
+   * Set table/region name of store
+   * 
+   * @param tableName
+   */
+  public void setTableName(String tableName);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java:45]: This is very specific to an implementation.. I don't think this is required in API

File: library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+   */
+  public void remove(Object key);
+
+  /**
+   * Get all the keys associated with key
+   * 
+   * @param key
+   * @return the list of all associated keys
+   */
+  public List<String> getKeys(Object key);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java:38]: What is the usecase for this? Can you please give example how will it work?

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+   * Generates key from operator id and window id to store unique operator
+   * checkpoints
+   * 
+   * @param operatorId
+   * @param windowId
+   * @return unique key for store
+   */
+  public static String generateKey(int operatorId, long windowId)
+  {
+    return String.valueOf(operatorId) + CHECKPOINT_KEY_SEPARATOR + String.valueOf(windowId);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java:100]: Should application Id not be used to generate this key? What happens if there are multiple applications running and they have operators with same operatorId?

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+   * 
+   */
+  @Override
+  public void save(Object object, int operatorId, long windowId) throws IOException
+  {
+
+    try {
+      store(generateKey(operatorId, windowId), object);
+      logger.info("saved check point object key {} region {}", generateKey(operatorId, windowId), applicationId);
+    } catch (Throwable t) {
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java]: why catch Throwable?

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+  {
+
+    if (!getStore().isConnected()) {
+      getStore().connect();
+    }
+
+    try {
+      store.remove(generateKey(operatorId, windowId));
+      logger.info("deleted object from store key {} region {}", generateKey(operatorId, windowId));
+    } catch (Throwable t) {
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java]: Why catch Throwable?

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+ * 
+ * @param <S>
+ *          Store implementation
+ */
+public abstract class AbstractKeyValueStorageAgent<S extends StorageAgentKeyValueStore>
+    implements AppIdAwareStorageAgent, Serializable
+{
+
+  protected S store;
+  protected String applicationId;
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java:47]: should applicationId be final?

Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java:47]: This should reflect the changes for ApplicationAwareStorageAgent in Apex, which it is not

File: library/src/main/java/com/datatorrent/lib/util/AppIdAwareStorageAgent.java (hunk 1 above)
Comment (ashishtadose) [inline library/src/main/java/com/datatorrent/lib/util/AppIdAwareStorageAgent.java]: This is required as ApplicationAwareStorageAgent is not released in Apex core and Malhar build takes core dependency of release candidates but not develop branches.

File: library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java (hunk 2 above)
Comment (ashishtadose) [inline library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java]: com.datatorrent.lib.db.KeyValueStore it contains additional methods such as getAll & putAll which are not relevant here also this requires one additional method which is not present there.

File: library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java (hunk 3 above)
Comment (ashishtadose) [inline library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java:45]: I thought it would be common for any key value store service as most of them have notion  of table, region to denote to particular storage instance.

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java (hunk 5 above)
Comment (ashishtadose) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java:100]: This would not happen as application ids are being used to create application specific tables or regions. 

For checkpointing of each application's operators a new table/region is created once the application is started.

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java (hunk 6 above)
Comment (ashishtadose) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java]: Initial though was, since this is abstract implementation and to ensure that any errors or exceptions thrown by underlying store should not cause disruption to application.

I can change this to catch to run time exception instead.

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java (hunk 8 above)
Comment (ashishtadose) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java:47]: Yes

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java (hunk 6 above)
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java]: You should not catch any exception.

File: library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java (hunk 4 above)
Comment (ashishtadose) [inline library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java:38]: Rationale for getKeys() method.

StorageAgent needs to provide below fours operations for operator checkpointing 
1 save(opertorId, windowId)
2 load(opertorId, windowId)
3 delete(opertorId, windowId)
4 getWindowIds(opertorId)

All Key value based systems perform operations based on key so to accommodate these operations need to form a key with combination of operatorId & windowId to uniquely identify checkpoints.
Note that each checkpoints will be written to different table/region by store implementation by using name provided by setTableName method.

To support 4th operation of getting windows ids related given operator id store implementation will have to scan keys using some form of query and return related windowids.

Please check provided concrete implementation of GeodeStore

**Comment by ashishtadose** [comment]
@tushargosavi  @gauravgopi123 
Please review.

File: contrib/pom.xml [hunk 9]
```diff
@@ -623,5 +628,12 @@
       <artifactId>super-csv-joda</artifactId>
       <version>2.3.1</version>
     </dependency>
+    <dependency>
+      <groupId>com.gemstone.gemfire</groupId>
+      <artifactId>gemfire</artifactId>
+      <version>8.2.0</version>
+      <type>jar</type>
+      <scope>compile</scope>
```
Comment (gauravgopi123) [inline contrib/pom.xml]: Default scope is compile so not required. I think the scope should be optional

Comment (gauravgopi123) [inline contrib/pom.xml]: I meant that this dependency should be set optional

Comment (ashishtadose) [inline contrib/pom.xml]: Yes dependency compile scope is default and should be optional. Fixed it.

File: contrib/pom.xml [hunk 10]
```diff
@@ -623,5 +628,11 @@
       <artifactId>super-csv-joda</artifactId>
       <version>2.3.1</version>
     </dependency>
+    <dependency>
+      <groupId>com.gemstone.gemfire</groupId>
+      <artifactId>gemfire</artifactId>
+      <version>8.2.0</version>
+      <type>jar</type>
```
Comment (gauravgopi123) [inline contrib/pom.xml]: please add `<optional>true</optional>`
`type` is by default jar so not required

**Comment by ashishtadose** [comment]
@gauravgopi123  Added optional tag to dependency.

File: pom.xml [hunk 11]
```diff
@@ -184,7 +184,6 @@
       <groupId>org.apache.apex</groupId>
       <artifactId>apex-engine</artifactId>
       <version>${apex.core.version}</version>
-      <scope>test</scope>
```
Comment (tweise) [inline pom.xml:195]: Why this change?

File: contrib/pom.xml [hunk 12]
```diff
@@ -53,6 +53,11 @@
         <updatePolicy>daily</updatePolicy>
       </releases>
     </repository>
+     <repository>
```
Comment (tweise) [inline contrib/pom.xml]: Is there no Geode release in Maven central?

File: pom.xml (hunk 11 above)
Comment (ashishtadose) [inline pom.xml:195]: This is required as AbstractKeyValueStorageAgent.java needs access to com.datatorrent.stram.plan.logical and test scope would cause compilation failure.

Please check - incubator-apex-malhar/library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java:[229,34]

File: contrib/pom.xml (hunk 12 above)
Comment (ashishtadose) [inline contrib/pom.xml]: No. Geode has not released a major release version yet.

File: library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java (hunk 2 above)
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/util/StorageAgentKeyValueStore.java]: Additional method can be added to extended Storage Agent. For the ones which are not required you can leave their implementation empty or throw exception

File: pom.xml (hunk 11 above)
Comment (tweise) [inline pom.xml:195]: Malhar code (except tests) should only depend on apex-api and apex-common, not engine internals.

Comment (ashishtadose) [inline pom.xml:195]: Hi @tweise 
Can you please take a look at incubator-apex-malhar/library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java:[229,34]

and suggest what can be done to accommodate this change without changing test score of apex-engine.

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+  }
+
+  /**
+   * Saves the yarn application id which can be used by create application
+   * specific table/region in KeyValue sore.
+   */
+  @Override
+  public void setApplicationAttributes(AttributeMap map)
+  {
+    this.applicationId = map.get(LogicalPlan.APPLICATION_ID);
```
Comment (tweise) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java]: Change this to DAGContext.APPLICATION_ID

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+   * @param windowId
+   *          window id of operator to checkpoint
+   */
+  @Override
+  public Object load(int operatorId, long windowId)
+  {
+    Object obj = null;
+    try {
+      obj = retrieve(generateKey(operatorId, windowId));
+      logger.info("retrieved object from store  key {} region {} ", generateKey(operatorId, windowId), applicationId);
```
Comment (tweise) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java]: change this to debug

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+  public void delete(int operatorId, long windowId) throws IOException
+  {
+
+    if (!getStore().isConnected()) {
+      getStore().connect();
+    }
+
+    try {
+      getStore().remove(generateKey(operatorId, windowId));
+      logger.info("deleted object from store key {} region {}", generateKey(operatorId, windowId));
```
Comment (tweise) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java]: debug

File: library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java (hunk 13 above)
Comment (ashishtadose) [inline library/src/main/java/com/datatorrent/lib/util/AbstractKeyValueStorageAgent.java]: Ok. will make these changes and push.

**Comment by ashishtadose** [comment]
@gauravgopi123 updated StorageAgentKeyValueStore to extend com.datatorrent.lib.db.KeyValueStore.
Can't completely remove it as to support Abstract implementation of storage agent which needs to integrate methods not supported in KeyValueStore. 
Also can't make these additional methods to throw exception as it is not supported in base interface.

@tweise @gauravgopi123  review comments have been incorporated.

File: contrib/pom.xml [hunk 16]
```diff
@@ -623,5 +628,11 @@
       <artifactId>super-csv-joda</artifactId>
       <version>2.3.1</version>
     </dependency>
+    <dependency>
+      <groupId>com.gemstone.gemfire</groupId>
```
Comment (markito) [inline contrib/pom.xml]: Would you guys like to point it to Geode instead of GemFire ?  For now it would need to be pulled from incubator repository and from nightlies but a release is just about to start. 

``` <dependency>
      <groupId>org.apache.geode</groupId>
      <artifactId>gemfire-core</artifactId>
      <version>${geode.version}</version>
</dependency>
```

Comment (tweise) [inline contrib/pom.xml]: That would be preferred. We would get the dependency from Apache snapshots,
right? In that case, no extra repository declaration should be required.

**Comment by ashishtadose** [comment]
During development Geode incubation branch wasn't stable so this was pointed other repo.
Recent Apache Geode alpha release is very stable.

Updated Geode dependency with Incubator Geode.

**Comment by ashishtadose** [comment]
@tweise can anyone look into branch conflicts.

**Comment by tweise** [comment]
@ashishtadose You need to rebase your branch and resolve all conflicts. You should also squash to commits. 

http://apex.incubator.apache.org/contributing.html#opening-pull-requests-contributors-

Doesn't this work depend on the upcoming Apex 3.3.0 release?

**Comment by ashishtadose** [comment]
Yes it depends on Apex 3.3.0 release. 
I will work on branch rebase & resolving conflicts.

**Comment by tweise** [comment]
@ashishtadose the 3.3.0 core dependency is in place now. Please squash and rebase.

**Comment by ashishtadose** [comment]
@tweise  Added test setup instructions, fixed logging & exceptions.
